### PR TITLE
Add optional cross-rotation for AWS and GCP

### PR DESCRIPTION
**Pull request by zkdev**
**What this PR does / why we need it**:
Adds optional "cross-rotation" (rotate User A with credentials of user B) for GCP and AWS ServiceAccounts.
Can be opt-in by adding attribute `rotation_cfg_name`.
AWS also requires to specify ServiceAccount name via attribute `iam_user_name`.

If `rotation_cfg_name` is not specified the regular "self-rotation" is performed.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by zkdev** [comment]
not yet tested E2E -> do not merge (yet)

**Comment by zkdev** [comment]
@ccwienk @8R0WNI3 adjusted as discussed, ready for review

File: cfg_mgmt/aws.py
Comment (8R0WNI3) [inline cfg_mgmt/aws.py:102]: maybe rather `... for deletion`?

File: cfg_mgmt/gcp.py
Comment (8R0WNI3) [inline cfg_mgmt/gcp.py:129]: same here

**Review by 8R0WNI3 — APPROVED** [review]
/lgtm

**Comment by zkdev** [comment]
@8R0WNI3 [flagged you as maintainer](https://github.com/gardener/cc-utils/commit/55016dd262e4e31be28445f283435d18b1cee9f2)

File: cfg_mgmt/aws.py
Comment (zkdev) [reply cfg_mgmt/aws.py:102]: copy-pasta ... , will adjust, thanks!

File: model/aws.py
Comment (ccwienk) [inline model/aws.py:34]: considering presense of `rotation_cfg_name` implies `iam_user_name` must also be set, how about adding this into validation. there are two options:

1. calculate required attributes in `_required_attributes` method
2. validate in `validate` method (see e.g. model/concourse.py for an example)

File: model/container_registry.py
Comment (ccwienk) [inline model/container_registry.py:86]: check for registry-type. if it is not GCR/AR, setting this attribute should at least emit a warning (or rather fail). Also, considering the cfg-name is of a different type (-> GCP I assume?), the type should also be configured, rather than being implied (implying type seems okay-ish if "rotation-cfg-element" is of same type as to-be-rotated element).

otoh: how about defining (-> model/base.py) a new type for rotation-cfg. it could be as simple as:
```
@dataclasses.dataclass
class CfgElementReference:
  type_name: str
  element_name: str
  purpose: str | None = None # might be useful for documenting in-place what the cfg-element-reference is used for
```

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: model/aws.py
Comment (zkdev) [reply model/aws.py:34]: Will add to required attributes, however we must ensure to explicitly validate in `aws.py`, as secret-replication handles validation-errors with warnings.

File: model/container_registry.py
Comment (zkdev) [reply model/container_registry.py:86]: This is a bit fuzzy as we have the generic cfg-type "container_registry" with its subclassification via "registry_type", but also the dedicated cfg-type "gcp".
However, both cfg-types use the very same rotation coding.

Therefore, if a cfg-element of type "container_registry" specifies a "rotation_cfg_name", it is implied that the referenced cfg is also of type "container_registry"
-> no need for explicit cfg-type specification

Comment (zkdev) [reply model/container_registry.py:86]: Thus, the only remaining value in introducing a `CfgElementReference` class would be the purpose attribute, however one could also simply inline that as Yaml comment.

Comment (zkdev) [reply model/container_registry.py:86]: > check for registry-type. if it is not GCR/AR, setting this attribute should at least emit a warning (or rather fail). 
good point, will add to validation

File: cfg_mgmt/util.py
Comment (ccwienk) [inline cfg_mgmt/util.py:435]: rather make this type-specific and explicit, please.

File: model/aws.py
Comment (ccwienk) [inline model/aws.py:31]: naming: while the typename is perfectly fine (CfgElementReference), the method's name is not ideal. it is about the cfg-element for rotation. the chosen name, however, drops the `element` part of the name (which reduces emphasis), and puts an emphesis on the aspect of the returned object being a a reference.

please rather name this:
- rotation_cfg_element
- rotation_cfg_element_reference
- rotation_cfg # this would be in sync with other namings of cfg-element-references we already have elsewhere

the same holds true for the attribute-name

File: model/aws.py
Comment (ccwienk) [inline model/aws.py:50]: it would be nice to document (preferably at the getters for rotation-cfg-element + iam_user_name) what those are intended for.

also, is it actually true that we always need a IAM user for credentials-rotation?

File: model/base.py
Comment (ccwienk) [inline model/base.py:188]: rather add this to cfg-factory as a convenience-method than here.

File: model/container_registry.py
Comment (ccwienk) [inline model/container_registry.py:98]: see comment to aws-cfg.

btw, if you _do_ want do deduplicate this, you might define a "CrossRotationMixin". If going for that approach, it would be fine to standardise on the name.

However, it is perfectly fine (and probably better, as we will have specific code for each element-type anyhow) to keep this type-specific. In which case, however, there should be no hardcoded assumptions as to the attribute-names.

File: model/gcp.py
Comment (ccwienk) [inline model/gcp.py:28]: ditto

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: model/aws.py
Comment (zkdev) [reply model/aws.py:50]: AWS cross-rotation **always** requires `iam_user_name`, yes.

As for the reasoning:
boto `iam_client` either implicitly uses its own iam-user (== self-rotation), or requires to explicitly give an user-name (== cross-rotation).
Btw, this holds also true for GCP rotation, however for GCP we [construct the user-name with ServiceAccount E-Mail and project-id](https://github.com/gardener/cc-utils/blob/3311905313346232547b52c8abebd0e24037b706/ccc/gcp.py#L72C5-L72C5).

**Review by ccwienk — APPROVED** [review]

**Pull request merged**
